rfcs/20211207-softmax-v2/README.md

```diff
@@ -0,0 +1,188 @@
+`data_desc` analogues will be replaced with `src_desc` correspondently.
+Implementation details will be similar to pooling_v2. We will also mark
+`softmax_v1` and `logsoftmax` API as deprecated, but in docs and release notes
+only. No code changes to not frustrate our users dealing with warning and
+potential errors (with DNNL_WERROR=ON).
+
+#### C API
+
+```c
+/// dnnl_types.h
+
+/// Kinds of primitives. Used to implement a way to extend the library with new
+/// primitives without changing the ABI.
+typedef enum {
+    ...
+    /// A softmax version 2 primitive (softmax with destination memory
+    /// descriptor and algorithm kind).
+    dnnl_softmax_v2,
+    ...
+} dnnl_primitive_kind_t;
+
+/// Kinds of algorithms.
+typedef enum {
+    ...
+    /// Softmax, numerically stable
+    dnnl_softmax_accurate,
+    /// Logsoftmax, numerically stable
+    dnnl_softmax_log,
+} dnnl_alg_kind_t;
+
+/// A descriptor of a Softmax operation.
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_softmax_v2.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training and
+    /// #dnnl_forward_inference, and #dnnl_backward_data.
+    dnnl_prop_kind_t prop_kind;
+    /// Source memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    /// Source gradient memory descriptor.
+    dnnl_memory_desc_t diff_src_desc;
+    /// The axis along which to perform the softmax.
+    int softmax_axis;
+    /// Softmax algorithm. Possible values: #dnnl_softmax_accurate and
+    /// #dnnl_softmax_log.
+    dnnl_alg_kind_t alg_kind;
+    /// Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+    /// Destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_dst_desc;
+} dnnl_softmax_v2_desc_t;
+
+/// Primitive descriptor query specification
+typedef enum {
+    ...
+    dnnl_query_softmax_v2_d, ///< softmax version 2 descriptor
+    ...
+} dnnl_query_t;
+```
+
+```c
+/// dnnl.h
+
+/// Initializes a descriptor for softmax v2 forward propagation primitive.
+///
+/// @param softmax_desc Output descriptor for a softmax primitive.
+/// @param prop_kind Propagation kind. Possible values are
+///     #dnnl_forward_training and #dnnl_forward_inference.
+/// @param alg_kind Softmax algorithm kind: either #dnnl_softmax_accurate, or
+///     #dnnl_softmax_log.
+/// @param src_desc Source memory descriptor.
+/// @param dst_desc Destination memory descriptor.
+/// @param softmax_axis Axis over which softmax is computed.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t DNNL_API dnnl_softmax_v2_forward_desc_init(
+        dnnl_softmax_v2_desc_t *softmax_desc, dnnl_prop_kind_t prop_kind,
+        dnnl_alg_kind_t alg_kind, const dnnl_memory_desc_t *src_desc,
+        const dnnl_memory_desc_t *dst_desc, int softmax_axis);
```

Comment:
We have both src_desc and dst_desc in the constructor now just because the data types in src and dst may be different, or there is any implication about memory layout? Previously in softmax v1, we only have data_dst which means src and dst should have the same data type and memory layout, right?

Reply:
Hi @TaoLv, thanks for the question. This assumption remains the same - the best performance may be expected if and only if dst_md has the same memory layout as source one. This may be achieved by either specifying the format directly if it is known for source and this is a preferred user way to initialize memory descriptors, or using tag::any for destination md and later on by querying dst_desc for format if this is the flow user follows. The user should expect dst_md will have the same memory format as src in this flow though. Hope it helps, thank you.

Reply:
Thank you for the explanation, @dzarukin. Another thing is the hint_fwd_pd in backward. I understand it's also required in softmax v1 hence not mentioned in this RFC. But what's the purpose of it? We need re-create hint_fwd_pd in the backward op. Previously, only data_desc is needed for hint_fwd_pd and we can create it from dst which is always needed in backward. But now, to create hint_fwd_pd, it seems we also need to accept src which is not present in backward op. Any suggestion for this case?

Reply:
The purpose of hint here is to be able to pass all memory descriptors with tag::any so that only tag for src will determine the whole flow:
src (defined X) -> dst (any ->X) ---bwd-(through-hint)---> dst (any -> X) -> diff_dst (any -> X) -> diff_src (any ->X).
You may use dst instead of src, just manually update data type to one from diff_src. This should be sufficient. Thank you.

Reply:
Thank you, @dzarukin!